attest/verifier/src/ias.rs

```diff
@@ -0,0 +1,244 @@
+        // trying to deprecate an existing authority in favor of a new one. In
+        // this case, they sign the old CA with the new CA, so things which
+        // trust the new CA also trust the old CA "for free". When everyone
+        // has the new CA in their trust list, they start issuing certs from
+        // the new CA, and stop renewing certs from the old CA. The old CA is
+        // gradually phased out of use as the certs it issued expire, and is
+        // eventually allowed to expire itself, or revoked by the new CA.
+        //
+        // As a result, if any pubkey in the actual chain matches the pubkey
+        // of a trust anchor, then we can consider the actual chain trusted.
+        //
+        // Lastly, it's possible that Intel provides multiple complete chains
+        // terminating at different root CAs. That is, the signature is tied
+        // to pubkey X, but there are multiple leaf certificates in the
+        // provided certs for pubkey X, and each one has its own path back to
+        // a trust anchor.
+
+        if report.chain.is_empty() {
+            return Err(Error::NoChain);
+        }
+
+        // Construct a verification profile for what kind of X509 chain we
+        // will support
+        let profile = Profile::new(
+            vec![HashType::Sha256, HashType::Sha384, HashType::Sha512],
+            vec![PkType::Rsa, PkType::Ecdsa],
+            vec![
+                EcGroupId::Curve25519,
+                EcGroupId::SecP256K1,
+                EcGroupId::SecP256R1,
+                EcGroupId::SecP384R1,
+                EcGroupId::SecP521R1,
+            ],
+            2048,
+        );
+
+        // Intel uses rsa-sha256 as their signature algorithm, which means
+        // the signature is actually over the sha256 hash of the data, not
+        // the data itself. mbedtls is primitive enough that we need to do
+        // these steps ourselves.
+        let hash = Sha256::digest(report.http_body.as_bytes());
+
+        // Cloned here because it needs to be mutable because mbedtls.
+        let mut trust_anchors = self.trust_anchors.clone();
+
+        let parsed_chain: Vec<MbedtlsBox<Certificate>> = report
+            .chain
+            .iter()
+            .filter_map(|maybe_der_bytes| Certificate::from_der(maybe_der_bytes).ok())
+            .collect();
+
+        parsed_chain
+            .iter()
+            // First, find any certs which contain the report signer's pubkey
+            .filter_map(|src_cert| {
+                let mut newcert = src_cert.clone();
+                newcert
+                    .public_key_mut()
+                    .verify(HashType::Sha256, hash.as_slice(), report.sig.as_ref())
+                    .and(Ok(newcert))
+                    .ok()
+            })
+            // Then construct a set of chains, one for each signer certificate
+            .filter_map(|cert| {
+                let mut signer_chain = vec![cert];
```

Comment:
This construct has been changed, in the original version, it was `let signer_chain: Vec<MbedtlsBox<Certificate>> = vec![cert];`, which is strange because cert here is of type `&MbedtlsBox<Certificate>`.

This should make the type `Vec<&MbedtlsBox<Certificate>>`, which feels like it would be closer to what we are looking for, but I'm very suspicious about all of this code and would appreciate a second set of eyes on it.

Reply:
I'm not sure this is right. I think that the previous filter map makes cert a MbedtlsBox<Certificate> rather than an &MbedtlsBox<Certificate> because it is cloned, which should return the object rather than a reference? maybe? Sorry if I'm misunderstanding.